### DISCUSSION TITLE
SSB / sneakernet

### DISCUSSION THREAD

**Discussion by samuk** [markqvist/Reticulum#22] [Ideas]
I'm interested in running SSB over reticulum https://scuttlebutt.nz/

It could enhance the store&forward capabilities and enable users to extend and link networks.

I'm imagining that at least some of the nodes in the network would run Linux  perhaps they are open hardware [Radxa zero's](https://wiki.radxa.com/Zero) with a [Lora shield.](https://github.com/reganhimself/LoRa-Hab-Hat) Perhaps [deployed by solar drone](https://youtu.be/aCM1DTj8wXI?t=45). These linux nodes could run something like [Kiwix hotspot](https://www.kiwix.org/en/downloads/kiwix-hotspot/) which among content from[ Rosettaphone](https://rosettaphone.org/) serves a [Manyverse.apk ](https://www.manyver.se/) over the wifi hotspot.

At deployment the devices would already contain the ~3Gb SSB database. [SSB over LoRa ](https://meshtastic.discourse.group/t/meshtastic-secure-scuttlebutt/3207/27)has already been done on Meshtastic.

As a user in a disaster zone I find a open wifi network, it captive portals me to install Manyverse. I install the apk and it syncs. I can send messages within Manyverse and whilst connected to the wifi it will send those over the LoRa network to other Radxa's and sync with those SSB databases.

I can additionally use Manyverse for direct peer<>peer comms over wifi or bluetooth, syncing back up with the Radxa/Lora network next time I connect to one of those nodes. I might have access to a motorbike and connect networks over long distances.

**Comment by 4c3e** [comment]
I'll throw my 2 cents in here because this is an area I'm particularly interested in.
Some of these ideas were discussed previously in [this](https://github.com/markqvist/Reticulum/discussions/15#discussioncomment-2279348) post by @markqvist and my following reply. Another relevant project in this area is [LibreRouter](https://www.netlab.tkk.fi/~jo/papers/2014-04-wons-liberouter.pdf).

SSB, Kiwix, and the other projects you linked are very interesting, thanks for linking them. SSB has a [fairly in-depth protocol](https://ssbc.github.io/scuttlebutt-protocol-guide/), which, from a quick glance, does not seem easily converted to Reticulum so it would probably be quite the development effort. That being said, the fundamental idea of "reticulum sneakernet transport" is possible right now (I think), although it has some minor UX issues.

Here's a sketch of a sneakernet network over reticulum:

- A group of users set up raspi hotspots across town, that are configured to be LXMF Propagation Nodes
- Users can install NomadNet or Sideband and run a local propagation node on their android/pc
- When Alice wants to send a message to Bob (messages in this case can be any arbitrary packet of information), but is unable to reach Bob, they can go to the nearby hotspot, lets call this "hotspot A", and sync their phone's propagation node with the hotspot A's propagation node. Now Alice has deposited their message into hotspot A.
- Now lets say a third user, Carl, is about to head across town, they first sync their phone with hotspot A, downloading Alice's encrypted message. Then Carl travels to Bob's part of town and syncs with the hotspot near Bob, call this "hotspot B".
- When Bob syncs with hotspot B, NomadNet/Sideband will download Alice's message for them and they will be able to decrypt and read it.

I haven't worked with LXMF Propagation nodes so far, just read about them, so this could be totally wrong. But from the description of LXMF Propagation nodes, I believe this system would work. The additional benefit here is that each raspi hotspot can have additional transports like TCP, i2p, or LoRa. "Sneakernet" would just be an additional transport layer added on top.

The UX issue I alluded to earlier would be that all the users would have to manually add the hotspots into their Reticulum config file as peers. They couldn't just simply walk up to a brand new LXMF node, jump on the hotspot wifi, and sync their messages. This UX issue is part of what I am trying to resolve with the [Transport Attached Sharable Links idea](https://github.com/markqvist/Reticulum/discussions/19).

**Reply by samuk** [reply]
I installed Piratebox on a OrangePi zero & documented it [here](https://github.com/Agroecology-Lab/edu-hotspot/blob/master/README.md) I'll have a go at installing an SSB pub server on it sometime. 

Once that's working I'll look again at the LoRa side. It seems it can be done on Meshtastic, so I may just end up using that, but I'd prefer Reticulum if possible.

**Comment by markqvist** [comment]
I've just had a look at Scuttlebutt, and quickly skimmed over the protocol documentation. I think it is perfectly possible to run the protocol over Reticulum directly, but as @4c3e also suggests, it will probably require some considerable effort to do this. That is *unless* the Scuttlebutt developers already has some sort of networking abstraction layer put in place, that easily allows using other transports than the TCP/IP with local discovery over UDP/IP they seem to be using right now. If that is the case, it might actually be pretty simple.

Another minimal way to do this would be to simply write a small Scuttlebutt peer utility that could tunnel Scuttlebutt traffic over an RNS network, and let all the connectivity between said peer and user devices such as phones and computers happen normally over TCP/IP. Either way is outside of what I have time to work on currently, so I will just leave those ideas here if anyone else wants to take them up and explore. I think it would be pretty cool and useful. I'd also be happy to provide info and guidance to any developers from the Scuttlebutt project if they want to add Reticulum support of some sort.

Regarding general sneakernet-type transport support, @4c3e is more or less spot on in that description. It is currently totally possible to do so technically, it just needs to be streamlined UX wise and documented so people can understand how to use it. Another way could be to install a Raspberry Pi with an LXMF propagation node running on it, and configure it to connect to some publicly open WiFi SSID, say "LXMF Distribution". Every time it came across such a network, it would just connect to it, and if any LXMF propagation nodes where present, sync with their message stores. By tuning announce times and sync intervals, this could work in practice while driving around a city or between towns, or just a couple of friends houses. Whatever really, you get the point :)

But yeah, the system needs to be improved and formalised a little to be more usable, and a few tweaks are also needed in the LXMF implementation to make it more reliable. Those are things I am currently focusing on.

**Comment by markqvist** [comment]
I've been messing around a bit more with SSB now, and really like the concept of a such a platform implemented entirely as a protocol. Very interesting, and seems very well executed. @samuk and @4c3e, does any of you have any recommendations of pub servers to join or have invites? Currently I am rather unconnected on my Manyverse install ;)

**Comment by samuk** [comment]
You could try one of the ones listed at the [top of the pub list ](https://github.com/ssbc/ssb-server/wiki/Pub-Servers)
eg

FreeSocial | @Jarland | pub.freesocial.co:8008:@ofYKOy2p9wsaxV73GqgOyh6C6nRGFM5FyciQyxwBd6A=.ed25519~ye9Z808S3KPQsV0MWr1HL0/Sh8boSEwW+ZK+8x85u9w=
-- | -- | --
Crypto Pub | Planetary.social | "crypto.planetary.pub:8008:@CBaNGA3MPZHPKo4gvrZsTCCV3DUgydmPnwK0UDRRbqQ=.ed25519~sGQ1aP56Kq5XhzI8KK7njn+Dexxggr3IVRt0ro5v7NA="

**Comment by markqvist** [comment]
Thanks @samuk, I will give it a try :)

**Reply by samuk** [reply]
If you make it in it's traditional to do a post with the #newpeople tag. I'll see you there.

As an aside I spotted this project today: https://berty.tech/docs/protocol/ not sure it offers anything beyond features Reticulum already has, but thought I'd mention it:

"The Berty Protocol provides secure communication between devices owned by the same account, communication between contacts in one-to-one conversations, as well as communication between several users in multi-member groups. This paper will explain how those points are implemented in a distributed and asynchronous way, both with or without internet access using IPFS and direct transports such as BLE. It will also describe how the Berty Protocol provides end-to-end encryption and perfect forward secrecy for all the exchanged messages."

**Reply by markqvist** [reply]
Thanks a bunch, that is good info to know :)

Interesting, I'll have a look at it.

**Comment by xloem** [comment]
I'm sorry for the off-topic comment, but did you ever find a separation between protocol and transport in ssb? Sometimes I look into starting a blockchain-based ssb pub.

**Reply by markqvist** [reply]
No, I have not looked that deeply into it yet.

**Comment by xloem** [comment]
Thank you. Where I left off I think I had figured it most effective to look at https://www.npmjs.com/package/secret-stack .

**Comment by markqvist** [comment]
Also, for anyone else lurking here, sneakernet and ssb-like functionality in LXMF is something I am actively working on, and will be out before too long :)
